### DISCUSSION TITLE
XMLStreamReader exceeds Quota maxElement limit of 5000

### DISCUSSION THREAD

**Discussion by paramdas11** [SalesforceCommerceCloud/sfcc-ci#542] [Team Posts]
We are using XMLStreamReader to read a large XML rating file imported from Bazaarvoice.
But the Job is failing due to com.demandware.core.quota.QuotaLimitExceededException: 
Limit for quota 'api.dw.io.XMLStreamReader.getXMLObject().maxElements' exceeded. Limit is 5000, actual is 5001

Is there any alternative approach to read the file ?

**Comment by adam-clason** [comment]
It seems you are not using the streaming capabilities of the XMLStreamReader class. I suspect you are reading almost the entire xml file with that getXMLObject() function and stuffing it into a variable. You will need to provide more code for context, but you want to loop over the individual nodes using the next() function, and only use the getXMLObject() function to pull small bits into variables as needed.

**Comment by pravngaur** [comment]
Hi Param,

I believe you would be getting this error as soon as the XMLSTREAMREADER
object is instantiated, it would be resulting in error.

In that case, you have to cut short the xml file so that it stays within
the permissible limit of nodes. Best way to do it is at the etl layer if
you have any. Else you would need to see what suits best in ur architecture.

But advise is not to do file churning in sfcc.

**Comment by paramdas11** [comment]
Hi Praveen
Thanks for you response. We are downloading the XML file from a SFTP 
location and then process it through a JOB controller and script files. 
Then we need to read the file to parse few data from XML elements.

During data reading through XMLStreamReader we get the exception.


The file is directly uploaded by our Review partner in the SFTP location, 
looks like can't split it.






Thanks & Regards,

Paramita Das
Application Architect-eCommerce


Phone: 91-33-400 97942 | Mobile: 91-9831032762
E-mail: paramdas@in.ibm.com
Chat: paramdas@in.ibm.com paramdas@gmail.com 
Find me on:  
"Life is not about waiting for the storms to pass, it's about learning how 
to dance in the rain" 


DLF IT Park, New Town
Kolkata, WB 700156
India

**Comment by GeneralPatten** [comment]
@paramdas11 — based on the quota error message, I suspect that you are likely trying to call XMLStreamReader.getXMLObject() on the top level document element. This would load the whole file into memory (generally not a good practice), so the quota ensures that this does not occur.

Instead, read through the XML stream, with a condition to match on the START_ELEMENT with the local name equal to the individual elements you will be pulling data from. Then call getXMLObject() on that element, and process as needed. In your case, maybe the feed is something like:

```
<feed>
  <products>
    <product>
    ...
    </product>
  </products>
</feed>
```

Then your script would be similar to

```
let fr = new FileReader(file, "UTF-8");
let xsr = new XMLStreamReader(fr);

 while (xsr.hasNext()) {
   if (xsr.next() === XMLStreamConstants.START_ELEMENT) {
     if (xsr.getLocalName() === "product") {
       // read single "product" as XML
       let productXml = xsr.getXMLObject();

       // process productXml
     }
   }
 }
 xsr.close();
 fr.close();
```

If individual product elements have more than 5000 children, then you would probably have to use local name checks for each individual element and grab the attribute and/or text values. 

See here for more info:
https://documentation.b2c.commercecloud.salesforce.com/DOC1/index.jsp?topic=%2Fcom.demandware.dochelp%2FDWAPI%2Fscriptapi%2Fhtml%2Fapi%2Fclass_dw_io_XMLStreamReader.html

**Comment by pravngaur** [comment]
I am assuming that you have validated and the script error outs at the
object instantiation only.

Again, in this case, file splitting has to happen anyhow. Should not be
done in sfcc. If your partner is not willing to do it then do it thru etl.

And if the script error outs at the point where instantiation is already
done then have a look at the advised code snippet on this thread where you
first check the nodes before accessing it.

**Comment by GeneralPatten** [comment]
@pravngaur – it does not need to be split. It just needs to be read as an XML stream as opposed to being loaded into memory. Very standard, very efficient, fastest execution  for any large files, regardless of platform. 

How do you think large catalog XML files get imported into and exported from SFCC? This is exactly what is happening when running import/export on the instances. It’s whats happening when data is deployed to development and production instances.

**Comment by paramdas11** [comment]
Hi Todd, Many thanks for your response !

However this is the full feed file which we are trying to import and read 
for first time. There is the existing code already which parses the delta 
feed file efficiently. 

So here the main concern is not parsing an XML file rather a giant XML 
file, which is 648157 KB vs the delta feed file which is around 700 KB and 
already being parsed.

Our Feed file format is as below -

<xml>
<feed>
<Product id=1>.......... 
<OverallRatingRange>5</OverallRatingRange><RatingsOnlyReviewCount>0</RatingsOnlyReviewCount><TotalReviewCount>0</TotalReviewCount>........</Product>
<Product id=2>.......... 
<OverallRatingRange>5</OverallRatingRange><RatingsOnlyReviewCount>0</RatingsOnlyReviewCount><TotalReviewCount>0</TotalReviewCount>........</Product>
...................................
...............................
<Product id=5000>.......... 
<OverallRatingRange>5</OverallRatingRange><RatingsOnlyReviewCount>0</RatingsOnlyReviewCount><TotalReviewCount>0</TotalReviewCount>........</Product>
</feed>

We are already parsing the product_id and rating XML elements by below 
code-

while(xmlReader.hasNext()) {
 
                                if (xmlReader.next() == XMLStreamCons
tants.START_ELEMENT && xmlReader.getLocalName() == 'Product') {
                                        productXML = 
xmlReader.readXMLObject();
                                        id = '';
                                        bvAverageRating = '';
                                        bvReviewCount = '';
                                        bvRatingRange = '';
 
                                        var ns = productXML.namespace();
                                        if(!empty(
productXML.ns::ExternalId.toString())) {
                                                id = 
productXML.ns::ExternalId.toString();
 
                                                if 
(!empty(productXML.ns::ReviewStatistics.ns::AverageOverallRat
ing.toString())) {
                                                    bvAverageRating = 
productXML.ns::ReviewStatistics.ns::AverageOverallRating.toString();
                                                }
                                                if 
(!empty(productXML.ns::ReviewStatistics.ns::TotalReviewCount.toString())) 
{
                                                    bvReviewCount = 
productXML.ns::ReviewStatistics.ns::TotalReviewCount.toString();
                                                }
                                                if 
(!empty(productXML.ns::ReviewStatistics.ns::OverallRatingRang
e.toString())) {
                                                    bvRatingRange = 
productXML.ns::ReviewStatistics.ns::OverallRatingRange.toString();
                                                }
 
                                                var product = 
ProductMgr.getProduct(BVHelper.decodeId(id));
                                                if (product != null) {
 product.custom.bvAverageRating = bvAverageRating;
 product.custom.bvReviewCount = bvReviewCount;
         product.custom.bvRatingRange = bvRatingRange;
         
 productIds.push(product.ID);
                                                    iProcessed++;
                                                } else {
                                                    iFailed++;
                                                }
                                        }
                                }
                    }
 
                        xmlReader.close();
                        fileReader.close();



Thanks & Regards,

Paramita Das
Application Architect-eCommerce


Phone: 91-33-400 97942 | Mobile: 91-9831032762
E-mail: paramdas@in.ibm.com
Chat: paramdas@in.ibm.com paramdas@gmail.com 
Find me on:  
"Life is not about waiting for the storms to pass, it's about learning how 
to dance in the rain" 


DLF IT Park, New Town
Kolkata, WB 700156
India

**Comment by pravngaur** [comment]
@todd dont get me wrong, please read through my response, i am still
requesting to validate where exactly the script errors out, if at the time
of reader object instantiation then only the split is suggested otherwise i
have mentioned to follow the code snippets provided by other community
members as for the same reason to check first for the child node type and
then only load the chunk.

Also, the deployment to prod and development is different that's
replication, nothing has to do with xml parsing.

@param please check on the todd's response.

**Comment by GeneralPatten** [comment]
@paramdas11 – What do the ellipsis in the <Product> element represent? Is it possible that a single <Product> element may have 5000+ descendants within it? I'm very familiar with the bv import code. The only place this exception can be thrown is on the line with `productXML = xmlReader.getXMLObject()`. Which indicates that it's not a file size issue, but an issue with one or more <product> elements having > 5000 descendants. In this case, you will need to stick with using the XMLStreamReader only to get the data. I would expect that the following psuedo code would work:

```
var fr = new FileReader(file, "UTF-8");
var xsr = new XMLStreamReader(fr);
var bvProduct;
var localName;

var localNames = [
  "ExternalId",
  "AverageOverallRating",
  "TotalReviewCount",
  "OverallRatingRange"
];

while (xsr.hasNext()) {
  if (xsr.next() === XMLStreamConstants.START_ELEMENT) {
    localName = xsr.getLocalName();
    if (localName === "product") {

      if (!empty(bvProduct)) {
        // Call function to update SFCC Product with bvProduct ExternalId, 
        // AverageOverallRating, TotalReviewCount, OverallRatingRange
      }
      bvProduct = {}; // initialize bvProduct to empty object
    }
    else if (~localNames.indexOf(localName)) {
      // if localName is found in localNames whitelist, store the value
      bvProduct[localName] = xsr.readElementText();
    }
  }
}

// Call function to update SFCC Product using last bvProduct object

xsr.close();
fr.close();
```
